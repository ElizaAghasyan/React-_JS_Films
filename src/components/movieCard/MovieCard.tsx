import { Link } from "react-router-dom";
import { apiConfig } from "../../config/apiConfig";
import { category } from "../../config/movieApi";
import { useRef, useState } from "react";
import TrailerModal from "../modal/TrailerModal";

import styles from './MovieCard.module.scss';
import PlayCircleFilledIcon from '@mui/icons-material/PlayCircleFilled';
import { makeStyles } from "@mui/styles";
import CloseIcon from "@mui/icons-material/Close";


const useStyles = makeStyles({
    play: {
        fontSize: '10rem',
        fill: '#3BB3DFFF',
        cursor: 'pointer'
    },
    close: {
        position: 'absolute',
        fontSize: '2.5em',
        top: '0.5em',
        left: '1em',
        color: '#fefefe',
        cursor: 'pointer'
    }
});

type movieCardProps = {
    item: {
        id: number;
        title: string;
        overview: string;
        poster_path: string;
        vote_average: number;
        backdrop_path: string;
        vote_count: number;
        name: string;
    }
}

const MovieCard = ({item}: movieCardProps) => {
    const [active, setActive] = useState<boolean>(false)
    const classes = useStyles();
    const detailsRef = useRef<HTMLIFrameElement | null>(null);
    const cardRef = useRef<HTMLIFrameElement | null>(null!);

    const link = '/' + category.movie + '/' + item?.id;

    const bg = apiConfig.w500(item?.poster_path || item?.backdrop_path);

    const setModalActive = () => setActive(!active);

    const handleClick = () => {
        if(detailsRef.current && detailsRef.current.classList.contains(styles.detailSide_active)) {
            detailsRef.current.classList.remove(styles.detailSide_active)
        }else {
            detailsRef.current && detailsRef.current.classList.add(styles.detailSide_active);
        }
    }

    return (
        <div ref={cardRef} className={styles.card}>
            <div className={`${styles.card_side_front} ${styles.card_side} `}>
                <div className={styles.card_pic} style={{backgroundImage: `url(${bg})`}}>
                </div>
                <Link to={link}>
                    <div className={styles.card_description}>
                        <div className={styles.card_description_header}>
                            <h3 className={styles.card_heading}>{item?.title || item?.name}</h3>
                            <p className={styles.card_genre}>Adventure Comedy Family</p>
                        </div>
                        <button className={styles.card_rating}>{item?.vote_average}</button>
                    </div>
                </Link>
            </div>
            <div ref={cardRef} className={`${styles.card_side_back} ${styles.card_side}`}>
                <PlayCircleFilledIcon onClick={setModalActive} className={classes.play}/>
                <h3 className={styles.card_watch}>Watch Now</h3>
                <button onClick={handleClick} className={styles.card_btn}>View Info</button>
            </div>
            <div ref={detailsRef} className={styles.detailSide}>
                <CloseIcon onClick={handleClick} className={classes.close}/>
                <div style={{backgroundImage: `url(${bg})`, height: '52rem'}}>
                    <div className={styles.content}>
                        <div className={styles.content_overlay}>
                            <h3>{item?.title || item?.name}</h3>
                            <p className={styles.card_genre}>Adventure Comedy Family</p>
                            <div className={styles.overview_style}>
                                <p className={styles.overview}>{item?.overview}</p>
                            </div>
                            <button onClick={setModalActive} className={`${styles.btn_overlay} ${styles.left}`}>Watch Now</button>
                            <Link to={link}>
                                <button onClick={setModalActive} className={`${styles.btn_overlay} ${styles.right}`}>See Details</button>
                            </Link>
                        </div>
                    </div>
                </div>
            </div>
            <TrailerModal active={active} id={item?.id}/>
        </div>
    );
}

export default MovieCard;
