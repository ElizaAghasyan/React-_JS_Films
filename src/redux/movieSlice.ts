import {createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import serverApi, {movieType} from "../config/movieApi";

const initialState = {
    movies: {},
}

interface MovieType  {
    id: number;
    title: string;
    overview: string;
    poster_path: string;
    vote_average: number;
    backdrop_path: string;
    vote_count: number;
    name: string;
}


export const fetchAsyncMovies = createAsyncThunk('movies/fetchAsyncMovies', async () => {
    const params = {page: 1}
    let response = await serverApi.getMovieList(movieType.popular, {params})
        .catch(err => {
            console.log(err)
        })
    let result = await response?.data.results[Math.floor(Math.random() * response.data.results.length - 1)]
    return result;
});


export const movieSlice = createSlice({
    name: 'movies',
    initialState,
    reducers: {
        addMovies: (state, {payload}) => {
            state.movies = payload
        }
    },
    extraReducers: (builder) => {
        builder.addCase(fetchAsyncMovies.pending, (state, action) => {
            console.log("Pending...")
        })
        builder.addCase(fetchAsyncMovies.fulfilled, (state: {}, action) => {
            return {
                ...state, movies: action.payload
            }
        })
        builder.addCase(fetchAsyncMovies.rejected, (state, action) => {
            console.log("Pending...")
        })
    }
});

export const {addMovies} = movieSlice.actions;

export const getBannerMovie = (state: { movies: { movies: MovieType }}) => state.movies.movies;


export default movieSlice.reducer;
