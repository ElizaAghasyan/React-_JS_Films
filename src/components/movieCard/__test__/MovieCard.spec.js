import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import MovieCard from '../MovieCard';
import {Router} from "react-router-dom";

describe('MovieCard Component', () => {

    it('MovieCard snapshot test', () => {
        const component = renderer.create(<Provider store={store}><Router><MovieCard /></Router></Provider>);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

});
