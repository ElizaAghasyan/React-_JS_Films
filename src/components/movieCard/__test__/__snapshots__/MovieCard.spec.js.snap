// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard Component MovieCard snapshot test 1`] = `
<div
  className="card"
>
  <div
    className="card_side_front card_side "
  >
    <div
      className="card_pic"
      style={
        Object {
          "backgroundImage": "url(https:://image.tmdb.org/t/p/w500/undefined)",
        }
      }
    />
    <a
      href="/movie/undefined"
      onClick={[Function]}
    >
      <div
        className="card_description"
      >
        <div
          className="card_description_header"
        >
          <h3
            className="card_heading"
          />
          <p
            className="card_genre"
          >
            Adventure Comedy Family
          </p>
        </div>
        <button
          className="card_rating"
        />
      </div>
    </a>
  </div>
  <div
    className="card_side_back card_side"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium makeStyles-play-1 css-i4bv87-MuiSvgIcon-root"
      data-testid="PlayCircleFilledIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 14.5v-9l6 4.5-6 4.5z"
      />
    </svg>
    <h3
      className="card_watch"
    >
      Watch Now
    </h3>
    <button
      className="card_btn"
      onClick={[Function]}
    >
      View Info
    </button>
  </div>
  <div
    className="detailSide"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium makeStyles-close-2 css-i4bv87-MuiSvgIcon-root"
      data-testid="CloseIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
    <div
      style={
        Object {
          "backgroundImage": "url(https:://image.tmdb.org/t/p/w500/undefined)",
          "height": "52rem",
        }
      }
    >
      <div
        className="content"
      >
        <div
          className="content_overlay"
        >
          <h3 />
          <p
            className="card_genre"
          >
            Adventure Comedy Family
          </p>
          <div
            className="overview_style"
          >
            <p
              className="overview"
            />
          </div>
          <button
            className="btn_overlay left"
            onClick={[Function]}
          >
            Watch Now
          </button>
          <a
            href="/movie/undefined"
            onClick={[Function]}
          >
            <button
              className="btn_overlay right"
              onClick={[Function]}
            >
              See Details
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
